ark/ops/ops_sendrecv_msll.cc

```diff
@@ -0,0 +1,206 @@
+         -1,
+         true} {}
+
+std::string MsllRecvOp::function_name(const OpConfig &) const {
+    Tensor *output = this->outputs[0];
+    CHECK(output->is_sequential());
+
+    int rank;
+    int src_rank;
+    this->args.get(&rank, 0);
+    this->args.get(&src_rank, 1);
+
+    return Op::function_name("ark::comm::recv_msll",
+                             {{
+                                 rank,      // Rank
+                                 src_rank,  // SrcRank
+                             }});
+}
+
+OpArgs MsllRecvOp::function_call_args(const OpConfig &) const { return {}; }
+
+MsllDeviceSyncOp::MsllDeviceSyncOp(const std::string &prec_type, Tensor *input,
+                                   Tensor *output, int nranks,
+                                   const std::string &name)
+    : Op{OP_DEVICE_SYNC_MSLL, prec_type, {input}, {output}, {{nranks}}, name,
+         &MsllSyncConfigMap,  -1,        true} {}
+
+std::string MsllDeviceSyncOp::function_name(const OpConfig &) const {
+    int nranks;
+    this->args.get(&nranks, 0);
+    return Op::function_name("ark::comm::device_sync_msll", {{nranks}});
+}
+
+OpArgs MsllDeviceSyncOp::function_call_args(const OpConfig &) const {
+    return {};
+}
+
+Tensor *Model::send_msll(Tensor *input, int sid, int dst_rank,
+                         std::size_t bytes, const std::string &name) {
+    size_t max_bytes = input->ldims_bytes();
+    if (max_bytes < bytes) {
+        LOG(ERROR, "invalid bytes: ", bytes, ", max: ", max_bytes);
+    }
+    if (bytes == 0) {
+        bytes = max_bytes;
+    }
+    LOG(DEBUG, "send_msll ", input->shape, " ", dst_rank, " ", bytes);
+    input->exported = true;
+
+    Tensor *recvbuf = this->tensor(input->shape, input->type);
+    recvbuf->imported_rank = dst_rank;
+    MsllSendOp op{"none",           input,    recvbuf, sid,
+                  this->impl->rank, dst_rank, bytes,   name};
+    return this->impl->add_op(op)[0];
+}
+
+Tensor *Model::send_done_msll(Tensor *input, int dst_rank,
+                              const std::string &name) {
+    LOG(DEBUG, "send_done_msll ", input->shape, " ", dst_rank);
+    MsllSendDoneOp op{"none", input, this->impl->rank, dst_rank, name};
+    return this->impl->add_op(op)[0];
+}
+
+Tensor *Model::recv_msll(int sid, int src_rank, size_t bytes, Tensor *output,
+                         const std::string &name) {
+    if (output == nullptr) {
+        if (bytes == 0) {
+            LOG(ERROR, "receive bytes cannot be 0");
+        }
+        output = this->tensor({DimType(bytes)}, BYTE);
+    }
+    output->exported = true;
+    size_t max_bytes = output->shape_bytes();
+    if (max_bytes < bytes) {
+        LOG(ERROR, "invalid bytes: ", bytes, ", max: ", max_bytes);
+    }
+    if (bytes == 0) {
+        bytes = max_bytes;
+    }
+    MsllRecvOp op{"none", output, sid, this->impl->rank, src_rank, bytes, name};
+    return this->impl->add_op(op)[0];
+}
+
+Tensor *Model::device_sync_msll(Tensor *input, int nranks,
```

Comment:
Seems we'd better have a separate file for this. We may also drop "msll" from the function name, as it's unrelated.

Reply:
Moved it to another file. This kernel is only provided by msll and depends on msll API.  So keep the msll in the function name.